--- ../../../q-e-qe-6.3MaX/PW/src/exx.f90	2018-06-30 01:21:56.000000000 +0900
+++ PW/src/exx.f90	2019-03-04 09:32:55.000000000 +0900
@@ -582,7 +582,11 @@
                    IF ( me_egrp == 0 ) THEN
 !$omp parallel do default(shared) private(ir) firstprivate(npol,nxxs)
                       DO ir=1,nxxs
+#if defined __CRAY 
+                         !DIR$ UNROLL 2
+#else
                          !DIR$ UNROLL_AND_JAM (2)
+#endif
                          DO ipol=1,npol
                             psic_all_nc(ir,ipol) = (0.0_DP, 0.0_DP)
                          ENDDO
@@ -590,7 +594,11 @@
 !$omp end parallel do
 !$omp parallel do default(shared) private(ir) firstprivate(npol,isym,nxxs) reduction(+:psic_all_nc)
                       DO ir=1,nxxs
+#if defined __CRAY 
+                         !DIR$ UNROLL 4
+#else
                          !DIR$ UNROLL_AND_JAM (4)
+#endif
                          DO ipol=1,npol
                             DO jpol=1,npol
                                psic_all_nc(ir,ipol) = psic_all_nc(ir,ipol) + &
@@ -607,7 +615,11 @@
 #else
 !$omp parallel do default(shared) private(ir) firstprivate(npol,nxxs)
                    DO ir=1,nxxs
+#if defined __CRAY
+                      !DIR$ UNROLL 2
+#else
                       !DIR$ UNROLL_AND_JAM (2)
+#endif
                       DO ipol=1,npol
                          psic_nc(ir,ipol) = (0._dp, 0._dp)
                       ENDDO
@@ -615,7 +627,11 @@
 !$omp end parallel do
 !$omp parallel do default(shared) private(ipol,jpol,ir) firstprivate(npol,isym,nxxs) reduction(+:psic_nc)
                    DO ir=1,nxxs
+#if defined __CRAY
+                      !DIR$ UNROLL 4
+#else
                       !DIR$ UNROLL_AND_JAM (4)
+#endif
                       DO ipol=1,npol
                          DO jpol=1,npol
                             psic_nc(ir,ipol) = psic_nc(ir,ipol) + conjg(d_spin(jpol,ipol,isym))* temppsic_nc(rir(ir,isym),jpol)
@@ -1389,7 +1405,11 @@
                                conjg(exxbuff(nrxxs+ir,jbnd-all_start(wegrp)+iexx_start,ikq))*temppsic_nc(ir,2,ii) )/omega
                          ENDDO
                       ELSE
+#if defined __CRAY
+!DIR$ LOOP_INFO cache_nt(rhoc)
+#else
 !DIR$ vector nontemporal (rhoc)
+#endif
                          DO ir = ir_start, ir_end
                             rhoc(ir,jbnd-jstart+1) = &
                                conjg(exxbuff(ir,jbnd-all_start(wegrp)+iexx_start,ikq))*temppsic(ir,ii) * omega_inv
@@ -1430,7 +1450,11 @@
                    DO jbnd=jstart, jend
                       ir_start = (irt - 1) * nblock + 1
                       ir_end = min(ir_start+nblock-1,nrxxs)
+#if defined __CRAY
+!DIR$ LOOP_INFO cache_nt(vc)
+#else
 !DIR$ vector nontemporal (vc)
+#endif
                       DO ir = ir_start, ir_end
                          vc(ir,jbnd-jstart+1) = facb(ir) * rhoc(ir,jbnd-jstart+1)*&
                                                 x_occupation(jbnd,ik) * nqs_inv
@@ -2265,7 +2289,11 @@
                       DO ibnd=ibnd_inner_start, ibnd_inner_end
                          ir_start = (irt - 1) * nblock + 1
                          ir_end = min(ir_start+nblock-1,nrxxs)
+#if defined __CRAY
+!DIR$ LOOP_INFO cache_nt(rhoc) 
+#else
 !DIR$ vector nontemporal (rhoc)
+#endif
                          DO ir = ir_start, ir_end
                             rhoc(ir,ibnd-ibnd_inner_start+1) = omega_inv * &
                               conjg(exxtemp(ir,ibnd-jblock_start+1)) * &
